llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6100,6 +6100,95 @@ bool InstCombinerImpl::replacedSelectWithOperand(SelectInst *SI,
   return false;
 }
 
+// Try to "strengthen" the RHS of compare based on known bits.
+// For example, replace `icmp ugt %x, 14` with `icmp ugt %x, 15` when
+// it is known that the two least significant bits of `%x` is zero.
+static Instruction *strengthenICmpUsingKnownBits(ICmpInst &I,
+                                                 KnownBits Op0Known,
+                                                 KnownBits Op1Known,
+                                                 unsigned BitWidth) {
+  if (!BitWidth)
+    return nullptr;
+  if (!(Op1Known.isConstant() && Op0Known.Zero.isMask()))
+    return nullptr;
+
+  Value *Op0 = I.getOperand(0);
+  ICmpInst::Predicate Pred = I.getPredicate();
+  Type *Ty = Op0->getType();
+  APInt RHSConst = Op1Known.getConstant();
+  bool TrueIfSigned = false;
+  // Don't break the SignBitCheck pattern;
+  if (InstCombiner::isSignBitCheck(Pred, RHSConst, TrueIfSigned))
+    return nullptr;
+
```

Comment:
You can add a check here to avoid breaking the SPF.

Reply:
@dtcxzyw Added a check not to break any select patterns.

Reply:
> @dtcxzyw Added a check not to break any select patterns.

Tests?

Reply:
@dtcxzyw I just added test for select patterns and sign-check pattern. Everything works as intended. I tried your specific example and the pattern's wasn't broken. There must be something else happening in that code. Can you double check 
 that the degradation is still there? If yes, can you provide the reduced test case?

Reply:
> @dtcxzyw I just added test for select patterns and sign-check pattern. Everything works as intended. I tried your specific example and the pattern's wasn't broken. There must be something else happening in that code. Can you double check that the degradation is still there? If yes, can you provide the reduced test case?

```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE = constant i64 67108864

define ptr @_ZN6Assimp14StackAllocator8AllocateEm(ptr %this, i64 %0) {
entry:
  %mul = shl i64 %0, 1
  store i64 %mul, ptr %this, align 8
  %call = call ptr @_ZSt3minImERKT_S2_S2_(ptr %this, ptr @_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE)
  %1 = load i64, ptr %call, align 8
  store i64 %1, ptr %this, align 8
  ret ptr null
}

define ptr @_ZSt3minImERKT_S2_S2_(ptr %__a, ptr %__b) {
entry:
  %0 = load i64, ptr %__b, align 8
  %1 = load i64, ptr %__a, align 8
  %cmp = icmp ult i64 %0, %1
  %__b.__a = select i1 %cmp, ptr %__b, ptr %__a
  ret ptr %__b.__a
}
```

Baseline (-O3):
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE = local_unnamed_addr constant i64 67108864

; Function Attrs: mustprogress nofree nosync nounwind willreturn memory(argmem: write)
define noalias ptr @_ZN6Assimp14StackAllocator8AllocateEm(ptr nocapture writeonly %this, i64 %0) local_unnamed_addr #0 {
entry:
  %mul = shl i64 %0, 1
  %1 = tail call i64 @llvm.umin.i64(i64 %mul, i64 67108864)
  store i64 %1, ptr %this, align 8
  ret ptr null
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: read)
define ptr @_ZSt3minImERKT_S2_S2_(ptr readonly %__a, ptr readonly %__b) local_unnamed_addr #1 {
entry:
  %0 = load i64, ptr %__b, align 8
  %1 = load i64, ptr %__a, align 8
  %cmp = icmp ult i64 %0, %1
  %__b.__a = select i1 %cmp, ptr %__b, ptr %__a
  ret ptr %__b.__a
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i64 @llvm.umin.i64(i64, i64) #2

attributes #0 = { mustprogress nofree nosync nounwind willreturn memory(argmem: write) }
attributes #1 = { mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: read) }
attributes #2 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

After this patch:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE = local_unnamed_addr constant i64 67108864

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: write)
define noalias noundef ptr @_ZN6Assimp14StackAllocator8AllocateEm(ptr nocapture writeonly %this, i64 %0) local_unnamed_addr #0 {
entry:
  %mul = shl i64 %0, 1
  %cmp.i = icmp ugt i64 %mul, 67108865
  %1 = select i1 %cmp.i, i64 67108864, i64 %mul
  store i64 %1, ptr %this, align 8
  ret ptr null
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: read)
define ptr @_ZSt3minImERKT_S2_S2_(ptr readonly %__a, ptr readonly %__b) local_unnamed_addr #1 {
entry:
  %0 = load i64, ptr %__b, align 8
  %1 = load i64, ptr %__a, align 8
  %cmp = icmp ult i64 %0, %1
  %__b.__a = select i1 %cmp, ptr %__b, ptr %__a
  ret ptr %__b.__a
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: write) }
attributes #1 = { mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: read) }
```

It looks like a phase ordering problem :(

Reply:
After inlining the call `%call = call ptr @_ZSt3minImERKT_S2_S2_(ptr %this, ptr @_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE)` the code looks like this:

```
; *** IR Dump After InlinerPass on (_ZN6Assimp14StackAllocator8AllocateEm) ***
define ptr @_ZN6Assimp14StackAllocator8AllocateEm(ptr %this, i64 %0) local_unnamed_addr {
entry:
  %mul = shl i64 %0, 1
  store i64 %mul, ptr %this, align 8
  %1 = load i64, ptr %this, align 8
  %cmp.i = icmp ult i64 67108864, %1
  %__b.__a.i = select i1 %cmp.i, ptr @_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE, ptr %this
  %2 = load i64, ptr %__b.__a.i, align 8
  store i64 %2, ptr %this, align 8
  ret ptr null
}
```

so the constant `_ZN6Assimp14StackAllocator18g_maxBytesPerBlockE` was replaced with its value in `icmp` instruction, but not in the `select`. Why?

Reply:
I see, the `select` can only be simplified after inlining and combining it with `load`.